airgun/entities/subscription.py

```diff
@@ -107,6 +107,18 @@ def search(self, value):
         view = self.navigate_to(self, 'All')
         return view.search(value)
 
+    def filter_columns(self, columns=None):
+        """Filters column headers
+        :param columns: name of column header to filter by
+        :return: tuple of the name of the headers
+        """
+        view = self.navigate_to(self, 'All')
+        if columns is not None:
+            view.columns_filter_checkboxes.fill(columns)
+            self.browser.plugin.ensure_page_safe()
+            view = self.navigate_to(self, 'All')
```

Comment:
Is this `navigate_to` necessary? You are on `All` and as far as I can tell, interacting with columns filter can't navigate you out of that page. So it kind of looks like code that doesn't do anything useful.

I suppose you added this because this is React page and without it, your script was trying to continue too early (i.e. before table columns were changed). Would it make sense to identify which exact part of `navigate_to` stabilizes your script and add it here directly?

Reply:
@mirzal You are correct in that assumption!  I'm still trying to figure out which part of navigate_to is used, but unable to identify the exact part of it.  This still fails from time to time as is so I do need help in this.

Reply:
I think it would be one of two places:
* [`am_i_here`](https://github.com/SatelliteQE/airgun/blob/master/airgun/navigation.py#L57) which would call [`view is_displayed`](https://github.com/SatelliteQE/airgun/blob/master/airgun/views/subscription.py#L100)
* subscription's own [`post_navigate`](https://github.com/SatelliteQE/airgun/blob/master/airgun/entities/subscription.py#L152), which eventually calls view is_displayed as well.

But `is_displayed` waits for div with id `subscriptions-table`, which - once it appears on page - does not disappear. When changing columns, React will first remove all cells in table heading one by one, then add expected set of cells, and then remove entire body of table, only to add it back (this time with new columns included). I'd say calling `self.navigate_to` doesn't actually wait for React to finish, it just delays further execution for long enough that React completes in the meantime.

As far as I know, there is no nice and clean solution. React doesn't expose internal state, so we have no way of knowing if it does something (and we should wait) or not.

One idea that comes to my mind is waiting for first row in table to have exact number of cells (which you should be able to compute based on number of columns before any action and results of changes from `checkbox.fill()`). This won't catch the case when you add and remove the same number of columns in one transaction. But I think you only care about displaying `virt-who` column, so it should suffice for now.

Reply:
I've tried those implementations before and they don't work as well as `navigate_to.`  I'm leaving it as is for now since it's working more consistently every time I've run it.

Reply:
Thanks for looking into this and trying other options as well. It's a bit unfortunate that they didn't work as well as navigation, but I take automation stability over everything else, so I'm OK with leaving it as is.